src/redis-cli.c

```diff
@@ -2025,6 +2052,16 @@ static sds *getSdsArrayFromArgv(int argc, char **argv, int quoted) {
 static int issueCommandRepeat(int argc, char **argv, long repeat) {
     while (1) {
         config.cluster_reissue_command = 0;
+        if (config.cluster_send_asking) {
+            if (cliSendAsking() != REDIS_OK) {
+                if (cliConnect(CC_FORCE) != REDIS_OK) {
+                    return REDIS_ERR;
+                }
+                if (cliSendAsking() != REDIS_OK) {
+                    return REDIS_ERR;
+                }
+            }
+        }
```

Comment:
I suppose this is better, but I feel that there are still too many calls to `cliConnect` (and also `cliSendCommand` and `cliSendAsking`) calls in this function.
some ignore errors, some add sleep, and some actually return an error (which is probably breaking #8884?).
i didn't think too deeply on that, or debug the code (busy with other things), but i have a strong feeling something is wrong.

i'd imagine this is the right approach:
- a state machine in a `while(1)` loop
- one iteration of the loop can run successfully though all the states (reconnect, asking, send command) and reach a break / return at the end.
- according to the state, some some steps may be skipped (i.e. don't reconnect or send asking if not needed).
- if anything fails we `continue` and start the loop from the top.
- if we have to reconnect, we throttle the loop by sleeping before `continue`.

let me know if you think i'm wrong.

Reply:
In fact I'm really doubt with retry. If the connection is broken(long time TCP is closed), I think we can retry once. But if some other thing is wrong,  retry is helpless.
If use state machine, maybe a lot of code needs modification. But it will be more clearer.
BTW there are many bugs in redis-cli. I try to fix some. Like https://github.com/redis/redis/pull/8982 https://github.com/redis/redis/pull/8945.
The retry in `reconnectingRedisCommand` is also wrong. It just use `redisConnect`. I don't fix it now.

Reply:
we discussed the reconnect retries in 8884, the conclusion was that since redis-cli is an interactive tool, we just need to throttle the retries (or mainly the error prints), and the user can just Ctrl+C whenever he wants.
anyway, you agree this function needs major refactoring to include just one call to each of the functions it calls?

Reply:
Agree.  There are too many these calls which make code hard to understand.

Reply:
> the user can just Ctrl+C whenever he wants

redis-cli is not only an interactive tool. It's also used in shell scripts.

Reply:
right.. in this case it's not clear if we want a retry loop or an error.
we had a retry loop so far (for a long time, right?), maybe some scripts would want to pass an argument for the amount of retries, or set it to 0.
anyway, i think that default to one retry is wrong, and the "must" thing we had to do is throttling the rate of prints.

Reply:
I'm not sure what's best. Perhaps the loop can be refactored a bit without changing the behaviour in this PR. Changing number of retries can be a different PR?

Reply:
I look the code deeper. I really can't find the infinite loop. It just retries once. Am I miss something?

Reply:
Maybe we need a test case to verify how many retries it does. ;-)

Reply:
@zuiderkwast I mentioned in https://github.com/redis/redis/pull/8884

After read the code and debug. I found the reason of the infinite loop is not the retry.
```
-> /home/hzw/database/redis/bin $ kill -9 20112;./redis-cli -p 7000 -c get a
[1]  + 20112 killed     ../redis-server redis.conf
Could not connect to Redis at 127.0.0.1:7002: Connection refused
Could not connect to Redis at 127.0.0.1:7002: Connection refused
```

When use host, there are only two line.

When use unix socket:
```
2027         config.cluster_reissue_command = 0;    /* this line set cluster_reissue_command */
2028         if (cliSendCommand(argc,argv,repeat) != REDIS_OK) {
2029             cliConnect(CC_FORCE);  /* cliConnect will use hostsocket (because cluster_reissue_command unset), so reconnect to the origin node, then receive MOVED */
2037         }

2039         /* Issue the command again if we got redirected in cluster mode */
2040         if (config.cluster_mode && config.cluster_reissue_command) {
2041             /* If cliConnect fails, sleep for a while and try again. */
2042             if (cliConnect(CC_FORCE) != REDIS_OK)     /* connect new node failed */
2043                 sleep(1);
```

Reply:
@huangzhw sorry for my luck of attention to details (busy with other things).
let me see if i get this right.
1. 8884 (or any recent change since 6.2) didn't create or remove any infinite loop. in some scenarios we had a loop and in others we didn't, and nothing really changed in that respect.
2. All that 8884 did is only to add some prints, and throttle the loop in the case there was one.
3. In the current code we normally have just two attempts (two calls to cliSendCommand in issueCommandRepeat) in case there's a connection error (when the second one fails, we return with REDIS_ERR)
4. in the current code we have a loop of infinite attempts only in a MOVED redirect and the new node isn't reachable? (or is it also / only if we're using a unix socket rather than a tcp port?)

either way i think there may be two actions to take:
- we need to decide if we wanna always create a loop, never create a loop, or keep the existing behavior.
  the two arguments to consider IMHO are:
  - if the current behavior haven't changed for a long time, then we might wanna keep it
  - if one behavior is common and the other is only triggered in rare cases (e.g. unix socket?), maybe we wanna make it consistent with the common one.
- this function is too complex and have too many calls to each of the functions it calls.
  i think we wanna refactor it to the state machine approach i described on the top.

Reply:
@oranagra Yes, you got it right.
For 4, In  the current code we have a loop of infinite attempts only in a MOVED redirect when using a unix socket. Tcp socket will not cause infinite loop. I think this really is a bug. So I think we wanna keep it and fix the infinite loop. The simple fix is when using `-c`, we can't use unix socket, so we will no longer encounter related bugs. But if we don't want to do this, we should fix carefully.

Reply:
How to handle redirect received on Unix socket?
1. Forbid Unix socket when -c is used. This is not backwards compatible. Maybe someone is using -c even if it's not a cluster and that will break if we forbid Unix socket with -c.
2. Error "Redirect not supported on Unix socket", etc.
3. Use TCP socket after redirect on Unix socket.

3 is better I guess, but if it's more work, maybe 2 is good enough?